File: app/src/test/java/realinterviews/CountStringInstancesTest.java
package realinterviews;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

public class CountStringInstancesTest {
    
    @Test
    void given_2MatchingPattern_when_countIstances_then_returns2() {
        CountStringInstances countStringInstances = new CountStringInstances();
        String pattern = "abr";
        String findIn = "abracadabra";

        int count = countStringInstances.countInstances(pattern, findIn);
        assertEquals(2, count);
    }

    @Test
    void given_1MatchingPattern_when_countIstances_then_returns1() {
        CountStringInstances countStringInstances = new CountStringInstances();
        String pattern = "abr";
        String findIn = "abracadab";

        int count = countStringInstances.countInstances(pattern, findIn);
        assertEquals(1, count);
    }

    @Test
    void given_0MatchingPattern_when_countIstances_then_returns0() {
        CountStringInstances countStringInstances = new CountStringInstances();
        String pattern = "bat";
        String findIn = "abracadabra";

        int count = countStringInstances.countInstances(pattern, findIn);
        assertEquals(0, count);
    }
}

File: app/src/test/kotlin/interviewpractice/CountStringInstancesTest.kt
package interviewpractice;

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.nulls.shouldNotBeNull
import kotlin.test.Test
import kotlin.test.assertNotNull

public class CountStringInstancesTest : BehaviorSpec({

    given("2 pattern matches") {
        val countStringInstances = CountStringInstances()
        val pattern = "abr"
        val findIn = "abracadabra"

        `when`("countInstances") {
            val count = countStringInstances.countStringInstances(pattern, findIn)

            then("return 2") {
                count.shouldBe(2)
            }
        }
    }

    given("1 pattern match") {
        val countStringInstances = CountStringInstances()
        val pattern = "abr"
        val findIn = "abracadab"

        `when`("countInstances") {
            val count = countStringInstances.countStringInstances(pattern, findIn)

            then("return 1") {
                count.shouldBe(1)
            }
        }
    }

    given("0 pattern match") {
        val countStringInstances = CountStringInstances()
        val pattern = "bat"
        val findIn = "abracadabra"

        `when`("countInstances") {
            val count = countStringInstances.countStringInstances(pattern, findIn)

            then("return 0") {
                count.shouldBe(0)
            }
        }
    }

})